benchmark/mha_block.py

```diff
@@ -0,0 +1,75 @@
+import torch
+from torchtext.modules import InProjContainer, MultiheadAttentionContainer, ScaledDotProduct
+from torch.nn.functional import multi_head_attention_forward as mha_forward
+import time
+
+
+def benchmark_mha_block():
+
+    def _run_benchmark(embed_dim, nhead, bsz, device, tgt_len, src_len=None):
+        # Build torchtext MultiheadAttention module
+        in_proj_container = InProjContainer(torch.nn.Linear(embed_dim, embed_dim),
+                                            torch.nn.Linear(embed_dim, embed_dim),
+                                            torch.nn.Linear(embed_dim, embed_dim))
+        MHA = MultiheadAttentionContainer(nhead, in_proj_container,
+                                          ScaledDotProduct(),
+                                          torch.nn.Linear(embed_dim, embed_dim)).to(device)
+
+        query = torch.rand((tgt_len, bsz, embed_dim)).to(device)
+        if src_len is None:
+            key = value = query
+            src_len = tgt_len
+        else:
+            key = value = torch.rand((src_len, bsz, embed_dim)).to(device)
+        attn_mask_2D = torch.randint(0, 2, (tgt_len, src_len)).to(torch.bool).to(device)
+        bias_k = bias_v = torch.rand((1, 1, embed_dim)).to(device)
+        print("starting torchtext.modules.MultiheadAttentionContainer")
+        t0 = time.monotonic()
+        for _ in range(100):
+            mha_output, attn_weights = MHA(query, key, value,
+                                           attn_mask=torch.stack([attn_mask_2D] * (bsz * nhead)),
+                                           bias_k=bias_k.repeat(1, bsz, 1).reshape(1, bsz * nhead, -1),
+                                           bias_v=bias_v.repeat(1, bsz, 1).reshape(1, bsz * nhead, -1))
+        print(time.monotonic() - t0)
```

Comment:
If you are benchmarking with CUDA, you need to add a `torch.cuda.synchronize()` before and after measuring the time, otherwise the timings won't be correct

Reply:
Thanks. Will add them there.

Reply:
The reason for this is that calls into cuda verions of operations are launched asynchronously. Only when you print a Tensor or convert it onto CPU can you be sure all operations have finished. Using synchronize here helps you make sure indeed all the work has finished and you're timing things correctly. Also see [torch.cuda](https://pytorch.org/docs/master/cuda.html).

Reply:
@zhangguanheng66 Could you share with us how your implementation performs compared to the PyTorch one after you have fixed the timing? Thanks.